core/market/src/negotiation/provider.rs

```diff
@@ -212,53 +219,173 @@ impl ProviderBroker {
     ) -> Result<ApprovalResult, AgreementError> {
         let dao = self.common.db.as_dao::<AgreementDao>();
         let agreement = {
-            self.common
-                .agreement_lock
-                .get_lock(&agreement_id)
-                .await
-                .lock()
-                .await;
+            let _hold = self.common.agreement_lock.lock(&agreement_id).await;
 
-            let agreement = match dao
-                .select(agreement_id, None, Utc::now().naive_utc())
+            let agreement = dao
+                .select(agreement_id, Some(id.identity), Utc::now().naive_utc())
                 .await
                 .map_err(|e| AgreementError::Get(agreement_id.to_string(), e))?
-            {
-                None => Err(AgreementError::NotFound(agreement_id.to_string()))?,
-                Some(agreement) => agreement,
-            };
+                .ok_or(AgreementError::NotFound(agreement_id.to_string()))?;
 
-            validate_transition(&agreement, AgreementState::Approved)?;
+            validate_transition(&agreement, AgreementState::Approving)?;
 
-            // `db.update_state` must be invoked after successful approve_agreement
-            // TODO: if dao.approve fails, Provider and Requestor have inconsistent state.
-            self.api.approve_agreement(&agreement, timeout).await?;
-            dao.approve(agreement_id, &app_session_id)
-                .await
-                .map_err(|e| AgreementError::UpdateState(agreement_id.clone(), e))?;
+            // TODO: Sign Agreement.
+            let signature = "NoSignature".to_string();
+            let timestamp = Utc::now().naive_utc();
 
+            let agreement = dao
+                .approving(&agreement.id, &app_session_id, &signature, &timestamp)
+                .await
+                .map_err(|e| AgreementError::UpdateState(agreement.id.clone(), e))?;
+
+            if let Some(session) = app_session_id {
+                log::info!(
+                    "AppSession id [{}] set for Agreement [{}].",
+                    &session,
+                    &agreement.id
+                );
+            }
             agreement
         };
 
-        self.common.notify_agreement(&agreement).await;
+        // Listen to Agreements notification before we start sending message, because otherwise
+        // we can lose events.
```

Comment:
I do not see how we can lose events here

Reply:
If listener doesn't exist, channel doesn't collect notifications. We must create channel first and than send message, which can result in event